### DISCUSSION TITLE
Multiple Channels vs. One Channel + Server Side handling [GroupChat]

### DISCUSSION THREAD

**Discussion by vaibhav2808** [centrifugal/centrifugo#666] [General]
Which approach is more suitable for creating large private chat rooms which can have 10k+ members and each users can be part of multiple groups: having separate group channels or user-specific channels?

Approach 1: Separate Group Channels
In this approach, each group chat has its own channel, and users subscribe to the group channels they have joined.

Approach 2: User-Specific Channels
In this approach, each user has their own channel, and the server manages groups without dedicated group channels. Messages are sent to the user-specific channels of the group participants.

Which approach would be easier to scale. I looked at the blog in centrifugo which was following Approach 2. However i had doubts if it would cause issues in large groups.

Also one more thing i am confused about is how to decide whether i want to send the notification or not. In approach 2, one way to achieve this is check if there is any active user that is subscribed to the private user channel and if not then send a notification about the message. Is there any better way to achieve this?

**Comment by FZambia** [comment]
Hello, this was actually discussed many times in community rooms.

It's hard to give an exact answer without knowing full application details. But some thoughts:

* if you will have a list of all chats on some screen in the app - then the idea with per-chat channels won't scale well. You will have to subscribe to many channels including those not currently visible on the chat list screen. In this perspective individual user channels are better.
* If you won't have chat list screen – then having per-chat channels may give you advantage since you can avoid broadcasts to large number of users. But even if you don't have such screen from the start - it may be required in the future and require refactoring.
* For large group chats and using individual user channels you can use `broadcast` server API. It's quite efficient, you can expect milliseconds time for thousands of channels in request. But you may want to have some queueing before Centrifugo and broadcast in batches when dealing with chats with huge number of members. This should scale well.

Personally, I'd go with individual user channels for the chat-like app.

> Also one more thing i am confused about is how to decide whether i want to send the notification or not. 

You can send notifications no matter what and check whether you need to display it comparing with the local app state, or think about more complex strategy like sending acks to messages and having a process on the backend which sends notifications for messages without ack for some period. This is mentioned in [Centrifugo FAQ](https://centrifugal.dev/docs/faq#what-if-i-need-to-send-push-notifications-to-mobile-or-web-applications) BTW. Relying on presence is possible I believe – but keep in mind it may be racy, user may be online, but when you send real-time message user can be away already.

**Reply by FZambia** [reply]
Since then we got a chat/messenger tutorial located here: https://centrifugal.dev/docs/tutorial/intro 

It uses personal user channel approach, which seems more suitable for an app where user can receive messages from many chat rooms within a single screen. If the app has isolated chat rooms, i.e. always one chat room on page, and no need to receive updates from another chat rooms - then channel per chat room could be a more efficient and simpler to scale solution.

So it depends on the use case.
